packages/server-wallet/src/protocols/defund-channel.ts

```diff
@@ -0,0 +1,66 @@
+import {Logger} from 'pino';
+
+import {ChainServiceInterface} from '../chain-service';
+import {ChainServiceRequest} from '../models/chain-service-request';
+import {ChallengeStatus} from '../models/challenge-status';
+import {DBDefundChannelObjective} from '../models/objective';
+import {Store} from '../wallet/store';
+import {WalletResponse} from '../wallet/wallet-response';
+
+export class ChannelDefunder {
+  constructor(
+    private store: Store,
+    private chainService: ChainServiceInterface,
+    private logger: Logger,
+    private timingMetrics = false
+  ) {}
+  public static create(
+    store: Store,
+    chainService: ChainServiceInterface,
+    logger: Logger,
+    timingMetrics = false
+  ): ChannelDefunder {
+    return new ChannelDefunder(store, chainService, logger, timingMetrics);
+  }
+
+  public async crank(
+    objective: DBDefundChannelObjective,
+    _response: WalletResponse
+  ): Promise<void> {
+    const {targetChannelId: channelId} = objective.data;
+    await this.store.transaction(async tx => {
+      const channel = await this.store.getAndLockChannel(channelId, tx);
+
+      if (!channel) {
+        this.logger.error(`No channel found for channel id ${channelId}`);
+        await this.store.markObjectiveStatus(objective, 'failed', tx);
```

Comment:
Why don't objectives have failure reasons? It seems important for a wallet user to know why an objective failed.

Reply:
I thought a bit about this but decided it was outside the scope of this PR. [It seems like we're removing the `failed` and` succeeded` status anways?](https://github.com/statechannels/statechannels/pull/3109/files)

I think this flags a good point though about how we want to handle "failed" objectives longer term. Do we just throw an exception or log an entry? Or do we want to update the objective database?

Reply:
Can someone to explain how "failed" objectives are meant to be handled?

Reply:
> Can someone to explain how "failed" objectives are meant to be handled?

I don't think there's much you can do besides retry them? They might be useful for debugging but I feel like we could just log that data out.

Considering this is the only place we use them and t[hey're being removed soon](https://github.com/statechannels/statechannels/pull/3109/files), I'll just remove them.

Reply:
I don't really know that they should be removed.

Currently, if the channel manager calls createChannel and it fails because there's insufficient funding in the ledger channel, there's no way for the channel manager to find that out. That appears to be a hole in the wallet API?

Reply:
(in this case, it would seem that the channel manager might react by depositing in the ledger channel before trying again.)

I'm not suggesting you make any change in this pr

Reply:
Ah good point!

Perhaps a discussion or spike is warranted here